crates/test-programs/tests/wasi-tokio.rs

```diff
@@ -0,0 +1,304 @@
+        ]
+    }
+    #[cfg(all(unix, not(target_os = "mac_os")))]
+    {
+        &[
+            ("ERRNO_MODE_UNIX", "1"),
+            // cap-std-sync does not support the sync family of fdflags
+            ("NO_FDFLAGS_SYNC_SUPPORT", "1"),
+        ]
+    }
+    #[cfg(target_os = "macos")]
+    {
+        &[
+            ("ERRNO_MODE_MACOS", "1"),
+            // cap-std-sync does not support the sync family of fdflags
+            ("NO_FDFLAGS_SYNC_SUPPORT", "1"),
+        ]
+    }
+}
+
+async fn run(name: &str, inherit_stdio: bool) -> Result<()> {
+    let workspace = prepare_workspace(name)?;
+    let stdout = WritePipe::new_in_memory();
+    let stderr = WritePipe::new_in_memory();
+    let r = {
+        let mut linker = Linker::new(&ENGINE);
+        add_to_linker(&mut linker, |cx| cx)?;
+
+        // Create our wasi context.
+        // Additionally register any preopened directories if we have them.
+        let mut builder = WasiCtxBuilder::new();
+
+        if inherit_stdio {
+            builder = builder.inherit_stdio();
+        } else {
+            builder = builder
+                .stdout(Box::new(stdout.clone()))
+                .stderr(Box::new(stderr.clone()));
+        }
+        builder = builder.arg(name)?.arg(".")?;
+        println!("preopen: {:?}", workspace);
+        let preopen_dir =
+            cap_std::fs::Dir::open_ambient_dir(workspace.path(), cap_std::ambient_authority())?;
+        builder = builder.preopened_dir(preopen_dir, ".")?;
+        for (var, val) in test_suite_environment() {
+            builder = builder.env(var, val)?;
+        }
+
+        let mut store = Store::new(&ENGINE, builder.build());
+        let instance = linker
+            .instantiate_async(&mut store, &get_module(name))
+            .await?;
+        let start = instance.get_typed_func::<(), ()>(&mut store, "_start")?;
+        start.call_async(&mut store, ()).await?;
+        Ok(())
+    };
+
+    r.map_err(move |trap: anyhow::Error| {
+        let stdout = stdout
+            .try_into_inner()
+            .expect("sole ref to stdout")
+            .into_inner();
+        if !stdout.is_empty() {
+            println!("guest stdout:\n{}\n===", String::from_utf8_lossy(&stdout));
+        }
+        let stderr = stderr
+            .try_into_inner()
+            .expect("sole ref to stderr")
+            .into_inner();
+        if !stderr.is_empty() {
+            println!("guest stderr:\n{}\n===", String::from_utf8_lossy(&stderr));
+        }
+        trap.context(format!(
+            "error while testing wasi-tests {} with cap-std-sync",
+            name
+        ))
+    })?;
+    Ok(())
+}
+
+// Below here is mechanical: there should be one test for every binary in
+// wasi-tests. The only differences should be should_panic annotations for
+// tests which fail.
+#[test_log::test(tokio::test(flavor = "multi_thread"))]
+async fn big_random_buf() {
+    run("big_random_buf", true).await.unwrap()
+}
```

Comment:
One thing that this pattern is losing is that if a test is added there's no guarantee that a test is added here. Could that be fixed perhaps with the build script generating something like:

```rust
use self::big_random_buf as _;
```

in the generated code? That'll trigger a compile error if there's a test without a corresponding `#[test]`

Reply:
Great idea, I will do that!